table/block_based_table_builder.cc

```diff
@@ -101,83 +102,100 @@ bool GoodCompressionRatio(size_t compressed_size, size_t raw_size) {
+          GetCompressFormatForVersion(kLZ4Compression, format_version),
+          raw.data(), raw.size(), compressed_output);
     case kLZ4HCCompression:
-      if (LZ4HC_Compress(
-              compression_ctx,
-              GetCompressFormatForVersion(kLZ4HCCompression, format_version),
-              raw.data(), raw.size(), compressed_output) &&
-          GoodCompressionRatio(compressed_output->size(), raw.size())) {
-        return *compressed_output;
-      }
-      break;     // fall back to no compression.
+      return LZ4HC_Compress(
+          compression_ctx,
+          GetCompressFormatForVersion(kLZ4HCCompression, format_version),
+          raw.data(), raw.size(), compressed_output);
     case kXpressCompression:
-      if (XPRESS_Compress(raw.data(), raw.size(),
-          compressed_output) &&
-          GoodCompressionRatio(compressed_output->size(), raw.size())) {
-        return *compressed_output;
-      }
-      break;
+      return XPRESS_Compress(raw.data(), raw.size(), compressed_output);
     case kZSTD:
     case kZSTDNotFinalCompression:
-      if (ZSTD_Compress(compression_ctx, raw.data(), raw.size(),
-                        compressed_output) &&
-          GoodCompressionRatio(compressed_output->size(), raw.size())) {
-        return *compressed_output;
-      }
-      break;     // fall back to no compression.
-    default: {}  // Do not recognize this compression type
+      return ZSTD_Compress(compression_ctx, raw.data(), raw.size(),
+                           compressed_output);
+    default:
+      // Do not recognize this compression type
+      return false;
   }
+}
 
-  // Compression method is not supported, or not good compression ratio, so just
-  // fall back to uncompressed form.
+}  // namespace
+
+// format_version is the block format as defined in include/rocksdb/table.h
+Slice CompressBlock(const Slice& raw, const CompressionContext& compression_ctx,
+                    CompressionType* type, uint32_t format_version,
+                    bool do_sample,
+                    std::string* compressed_output,
+                    std::string* sampled_output_fast,
+                    std::string* sampled_output_slow) {
+  *type = compression_ctx.type();
+
+  if (compression_ctx.type() == kNoCompression &&
+      !compression_ctx.sample_for_compression()) {
+    return raw;
+  }
+
+  // If requested, we sample one in every N block with a
+  // fast and slow compression algorithm and report the stats.
+  // The users can use these stats to decide if it is worthwhile
+  // enabling compression and they also get a hint about which
+  // compression algorithm wil be beneficial.
+  if (do_sample && compression_ctx.sample_for_compression() &&
+      Random::GetTLSInstance()->OneIn(
+          (int)compression_ctx.sample_for_compression()) &&
+      sampled_output_fast && sampled_output_slow) {
+    // Sampling with a fast compression algorithm
+    if (LZ4_Supported() || Snappy_Supported()) {
```

Comment:
I am a little concerned about hard-coded compression algorithms here. If some one wants to sample other algorithms, does he need to add new code here?

Reply:
My previous diff (on internal phabricator) passed in a setting where one could specify the algorithm to use for sampling (just like the algorithm to use for compression). After some internal discussion (see the diff) we decided to sample with one fast and one slow algorithm.

In general, we are looking for an indication for compressibility of data and lz4+snappy on the fast end and Zstd+zlib on the high compression end covers the spectrum pretty well -- new compression algorithms don't really come around that often.